src/main/java/io/dockstore/provision/GSPlugin.java

```diff
@@ -0,0 +1,271 @@
+            } else {
+                // When Blob size is big or unknown use the blob's channel reader.
+                try (ReadChannel reader = blob.reader()) {
+                    WritableByteChannel channel = Channels.newChannel(writeTo);
+                    ByteBuffer bytes = ByteBuffer.allocate(DOWNLOAD_BUFFER_SIZE);
+                    try {
+                        int limit;
+                        long runningTotal = 0;
+                        while ((limit = reader.read(bytes)) > 0) {
+                            bytes.flip();
+                            channel.write(bytes);
+                            bytes.clear();
+                            runningTotal = runningTotal + limit;
+                            printer.handleProgress(runningTotal, inputSize);
+                        }
+                    } catch (IOException e) {
+                        System.err.println("Could not download file. IO exception:" + e.getMessage());
+                        return false;
+                    }
+
+                }
+            }
+
+            writeTo.close();
+            return true;
+        }
+
+        public boolean uploadTo(String destPath, Path sourceFile, Optional<String> metadata) {
+            long inputSize = sourceFile.toFile().length();
+
+            String contentType = "application/octet-stream";
+            try {
+                contentType = Files.probeContentType(sourceFile);
+            } catch (IOException e) {
+                System.err.println("Could not get content type of file. Using default content type. IO exception:" + e.getMessage());
+            }
+
+            // Bug on Mac where Files.probeContentType returns null
+            if (contentType == null)
+                contentType = "application/octet-stream";
+
+            Storage gsClient = getGoogleGSClient();
+
+            String trimmedPath = destPath.replace("gs://", "");
+            List<String> splitPathList = Lists.newArrayList(trimmedPath.split("/"));
+            String bucketName = splitPathList.remove(0);
+            String blobName = String.join(File.separator, splitPathList);
+
+            BlobId blobId = BlobId.of(bucketName, blobName);
+            BlobInfo blobInfo;
+
+
+            // To test metadata uncomment the lines below
+            //Map<String,String> myMap = new HashMap<String, String>();
+            //myMap.put("goat", "bleat");
+            //myMap.put("cow", "moo");
+            //metadata = Optional.of(myMap.toString());
+
+            if (metadata.isPresent()) {
+                Gson gson = new Gson();
+                Type type = new TypeToken<Map<String, String>>() {
+                }.getType();
+
+                try {
+                    Map<String, String> map = gson.fromJson(metadata.get(), type);
+                    for (Map.Entry<String, String> entry : map.entrySet()) {
+                        System.out.println("Loading " + entry.getKey() + "->" + entry.getValue());
```

Comment:
I don't think you should be printing this. Or is there a reason you are doing so beyond debugging?

Reply:
Commented out; should there be a logger?

Reply:
We don't think logging works, but there is a ticket to investigate, ga4gh/dockstore#1483

Reply:
I have left logging implementation for the other ticket